controllers/solrcloud_controller.go

```diff
@@ -960,80 +950,160 @@ func (r *SolrCloudReconciler) indexAndWatchForTLSSecret(mgr ctrl.Manager, ctrlBu
-			// capture the hash of the secret and stash in an annotation so that pods get restarted if the cert changes
-			if instance.Spec.SolrTLS.RestartOnTLSSecretUpdate {
-				if tlsCertBytes, ok := foundTLSSecret.Data[util.TLSCertKey]; ok {
-					tls.CertMd5 = fmt.Sprintf("%x", md5.Sum(tlsCertBytes))
-				} else {
-					return nil, fmt.Errorf("%s key not found in TLS secret %s, cannot watch for updates to"+
-						" the cert without this data but 'solrTLS.restartOnTLSSecretUpdate' is enabled",
-						util.TLSCertKey, foundTLSSecret.Name)
+			if serverCert.RestartOnTLSSecretUpdate {
+				tls.ServerConfig.CertMd5, err = r.computeCertMd5(foundTLSSecret)
+				if err != nil {
+					return nil, err
 				}
+				tls.ServerConfig.CertMd5Annotation = util.SolrTlsCertMd5Annotation
 			}
 
-			if _, ok := foundTLSSecret.Data[instance.Spec.SolrTLS.PKCS12Secret.Key]; !ok {
+			if _, ok := foundTLSSecret.Data[serverCert.PKCS12Secret.Key]; !ok {
 				// the keystore.p12 key is not in the TLS secret, indicating we need to create it using an initContainer
-				tls.NeedsPkcs12InitContainer = true
+				tls.ServerConfig.NeedsPkcs12InitContainer = true
 			}
 		}
 
-		if instance.Spec.SolrTLS.TrustStoreSecret != nil {
+		if serverCert.TrustStoreSecret != nil {
 			// verify the TrustStore secret is configured correctly
-			passwordSecret := instance.Spec.SolrTLS.TrustStorePasswordSecret
-			if passwordSecret == nil {
-				passwordSecret = instance.Spec.SolrTLS.KeyStorePasswordSecret
+			err = r.verifyTruststoreConfig(serverCert, instance.Namespace)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			// does the supplied keystore secret also contain the truststore?
+			if _, ok := foundTLSSecret.Data[util.DefaultPkcs12TruststoreFile]; ok {
+				// there's a truststore in the supplied TLS secret
+				serverCert.TrustStoreSecret = &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{Name: foundTLSSecret.Name},
+					Key:                  util.DefaultPkcs12TruststoreFile,
+				}
+				serverCert.TrustStorePasswordSecret = serverCert.KeyStorePasswordSecret
 			}
-			_, err := r.verifyTLSSecretConfig(instance.Spec.SolrTLS.TrustStoreSecret.Name, instance.Namespace, passwordSecret)
+		}
+
+		// is there a client TLS config too?
+		if tls.ClientConfig != nil {
+			if tls.ClientConfig.Options.PKCS12Secret == nil {
+				// cannot mix options with the client cert, if the server cert comes from a secret, so too must the client, not a mountedTLSDir
+				return nil, fmt.Errorf("invalid TLS config, the 'solrClientTLS.pkcs12Secret' option is required when using a secret for server cert")
+			}
+
+			clientOpts := tls.ClientConfig.Options
+			// shouldn't configure a client cert if it's the same as the server cert
+			if clientOpts.PKCS12Secret == tls.ServerConfig.Options.PKCS12Secret {
+				return nil, fmt.Errorf("invalid TLS config, the 'solrClientTLS.pkcs12Secret' option should not be the same as the 'solrTLS.pkcs12Secret'")
```

Comment:
Will this actually result in an error when running Solr?

Reply:
Probably not an issue for Solr but seemed like that would be a mis-configuration / accident by the user since it wouldn't be sensible to use the same cert for the client & server (you don't need a separate client in that case) so I felt raising an error would help draw attention to a mistake.